rust/cubesql/cubesql/src/sql/session_manager.rs

```diff
@@ -41,60 +50,72 @@ impl SessionManager {
         protocol: DatabaseProtocol,
         client_addr: String,
         client_port: u16,
-    ) -> Arc<Session> {
+        extra_id: Option<String>,
+    ) -> Result<Arc<Session>, CubeError> {
         let connection_id = self.last_id.fetch_add(1, Ordering::SeqCst);
 
-        let sess = Session {
+        let session_ref = Arc::new(Session {
             session_manager: self.clone(),
             server: self.server.clone(),
             state: Arc::new(SessionState::new(
                 connection_id,
+                extra_id.clone(),
                 client_addr,
                 client_port,
                 protocol,
                 None,
                 Duration::from_secs(self.server.config_obj.auth_expire_secs()),
                 Arc::downgrade(self),
             )),
-        };
-
-        let session_ref = Arc::new(sess);
+        });
 
         let mut guard = self.sessions.write().await;
 
-        guard.insert(connection_id, session_ref.clone());
+        if let Some(extra_id) = extra_id {
+            if guard.uid_to_session.contains_key(&extra_id) {
```

Comment:
AFAIU, `extra_id` is user-controlled, so this place can trigger hashing of strings of arbitrary length.
Do we want to limit it?